aten/src/ATen/native/cudnn/Conv_v8.cpp

```diff
@@ -76,99 +139,533 @@ void filterEngineConfigs(
-    auto plan = cudnn_frontend::ExecutionPlanBuilder()
-        .setHandle(handle)
-        .setEngineConfig(cfg)
-        .build();
-
-    auto workspace_size = plan.getWorkspaceSize();
-    auto workspace = at::empty({workspace_size}, input.options().dtype(kByte));
-    void *data_ptrs[] = {input.data_ptr(), output.data_ptr(), weight.data_ptr()};
-    // std::cout << plan.describe() << " requires workspace " << workspace_size << std::endl;
-    int64_t uids[] = {'x', 'y', 'w'};
-    auto variantPack = cudnn_frontend::VariantPackBuilder()
-        .setWorkspacePointer(workspace.data_ptr())
-        .setDataPointers(3, data_ptrs)
-        .setUids(3, uids)
-        .build();
-    AT_CUDNN_CHECK(cudnnBackendExecute(handle, plan.get_raw_desc(), variantPack.get_raw_desc()));
-  };
+};
 
-  auto search = engine_cache.find(key);
-  if (search != engine_cache.end()) {
-    run(search->second);
-    return;
-  }
+BenchmarkCache<cudnn_frontend::ExecutionPlan, CacheKey> benchmark_cache;
+BenchmarkCache<cudnn_frontend::ExecutionPlan, CacheKeyFused> benchmark_cache_fused;
+
+} // namespace
+void get_cachekey(CacheKey& key, const cudnnBackendDescriptorType_t operation, const Tensor& y, const Tensor& x, const Tensor& w, const IntArrayRef padding, const IntArrayRef stride, const IntArrayRef dilation, int64_t groups, bool deterministic, bool allow_tf32) {
+  memset(&key, 0, sizeof(key));
+  setConvolutionParams(&key.params, x, w, padding, stride, dilation, groups, deterministic, allow_tf32);
+  key.operation = operation;
+  key.x_alignment = getAlignment(x);
+  key.y_alignment = getAlignment(y);
+  key.w_alignment = getAlignment(w);
+}
+
+void get_cachekey_fused(CacheKeyFused& key, const Tensor& y, const Tensor& x, const Tensor& w, const Tensor& z, const Tensor& b, const float alpha, const IntArrayRef padding, const IntArrayRef stride, const IntArrayRef dilation, int64_t groups, bool deterministic, bool allow_tf32) {
+  memset(&key, 0, sizeof(key));
+  setConvolutionParams(&key.params, x, w, padding, stride, dilation, groups, deterministic, allow_tf32);
+  key.x_alignment = getAlignment(x);
+  key.y_alignment = getAlignment(y);
+  key.w_alignment = getAlignment(w);
+  key.z_alignment = getAlignment(z);
+  key.b_alignment = getAlignment(b);
+  key.alpha = alpha;
+}
 
-  auto op = cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_CONVOLUTION_FORWARD_DESCRIPTOR)
-      .setxDesc(getTensorDescriptor(input, 'x', key.input_alignment))
-      .setyDesc(getTensorDescriptor(output, 'y', key.output_alignment))
-      .setwDesc(getTensorDescriptor(weight, 'w', key.weight_alignment))
-      .setcDesc(getConvDescriptor(key.params.dataType, padding, stride, dilation))
+void run_conv_plan(cudnnHandle_t handle, const Tensor& x, const Tensor& y, const Tensor& w, const cudnn_frontend::ExecutionPlan& plan) {
+  auto workspace_size = plan.getWorkspaceSize();
+  Tensor workspace = at::empty({workspace_size}, x.options().dtype(kByte));
```

Comment:
you don't need to allocate workspace as an empty tensor, that incurs additional overhead. Better use `.allocate` API from caching allocator

Reply:
Would there need to be any extra checks/handling to ensure that the allocation would be on the same device as `x`?